iaso/api/storage.py

```diff
@@ -0,0 +1,334 @@
+    logs = StorageLogSerializer(many=True, source="log_entries")
+
+    class Meta(StorageSerializer.Meta):
+        fields = StorageSerializer.Meta.fields + ("logs",)
+
+
+class StorageViewSet(ListModelMixin, viewsets.GenericViewSet):
+    permission_classes = [permissions.IsAuthenticated, HasPermission("menupermissions.iaso_storages")]  # type: ignore
+    serializer_class = StorageSerializer
+
+    def get_queryset(self):
+        """Devices get filtered by account, and optionally by a few URL parameters"""
+
+        # In all cases, we'll only return results for the account of the user
+        qs = StorageDevice.objects.filter(account=self.request.user.iaso_profile.account)
+
+        # There are also a few optional URL filters
+        filter_status = self.request.query_params.get("status")
+        if filter_status is not None:
+            qs = qs.filter(status=filter_status)
+
+        filter_reason = self.request.query_params.get("reason")
+        if filter_reason is not None:
+            qs = qs.filter(status_reason=filter_reason)
+
+        filter_type = self.request.query_params.get("type")
+        if filter_type is not None:
+            qs = qs.filter(type__in=filter_type.split(","))
+
+        filter_search = self.request.query_params.get("search")
+        if filter_search is not None:
+            # TODO: implement: search on entity or storage id
+            pass
+
+        return qs
+
+    def list(self, request):
+        """
+        Endpoint used to list/search devices on the web interface
+
+        GET /api/storage/
+        """
+        limit_str = request.GET.get("limit", None)
+        page_offset = request.GET.get("page", 1)
+        order = request.GET.get("order", "updated_at,id").split(",")
+
+        queryset = self.get_queryset()
+        queryset = queryset.order_by(*order)
+        serializer = StorageSerializer
+
+        if limit_str is not None:
+            limit = int(limit_str)
+            page_offset = int(page_offset)
+            paginator = Paginator(queryset, limit)
+            res = {"count": paginator.count}
+            if page_offset > paginator.num_pages:
+                page_offset = paginator.num_pages
+            page = paginator.page(page_offset)
+            res["results"] = serializer(page.object_list, many=True).data
+            res["has_next"] = page.has_next()
+            res["has_previous"] = page.has_previous()
+            res["page"] = page_offset
+            res["pages"] = paginator.num_pages
+            res["limit"] = limit
+            return Response(res)
+        else:
+            return Response(StorageSerializer(queryset, many=True).data)
+
+    @action(detail=False, methods=["post"])
+    def blacklisted(self, request):
+        """
+        Endpoint used to blacklist a single device from the web interface
+
+        POST /api/storage/blacklisted/
+        """
+        # TODO: permissions: spec mentions "permission to modify storage", what does it mean exactly?
```

Comment:
I think we were considering adding a supplementary perm?

Reply:
I asked to the client and she said no...